### PR TITLE
Deregister connection pool metrics from Micrometer (or alternate registry) when disposing the connection pool

### PR DESCRIPTION
**Pull request by manolama**
Deregister pool metrics from Micrometer (or alternate registry) in order to release resources when there is a large churn of connection pools and endpoints.

### CODE REVIEW

File: reactor-netty-core/src/main/java/reactor/netty/resources/ConnectionProvider.java [hunk 1]
```diff
@@ -817,5 +817,7 @@ interface MeterRegistrar {
 
 		void registerMetrics(String poolName, String id, SocketAddress remoteAddress, ConnectionPoolMetrics metrics);
 
+		void deRegisterMetrics(String poolName, String id, SocketAddress remoteAddress);
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/resources/ConnectionProvider.java:820]: Let's make JAPICMP happy (I will remove the default implementation in Reactor Netty version 2.0.x where we can make this incompatible change)

```
Japicmp report was filtered and interpreted to find the following incompatibilities:
source incompatible change: **** MODIFIED INTERFACE: PUBLIC ABSTRACT STATIC reactor.netty.resources.ConnectionProvider$MeterRegistrar  (not serializable)
source incompatible change:     +++* NEW METHOD: PUBLIC(+) ABSTRACT(+) void deRegisterMetrics(java.lang.String, java.lang.String, java.net.SocketAddress)
```

```suggestion
		/**
		 * Invoked when a connection pool is disposed.
		 *
		 * @param poolName the pool name
		 * @param id the pool id
		 * @param remoteAddress the remote address
		 * @since 1.0.26
		 */
		default void deRegisterMetrics(String poolName, String id, SocketAddress remoteAddress) {
		}
```

File: reactor-netty-core/src/test/java/reactor/netty/resources/ConnectionProviderTest.java [hunk 2]
```diff
@@ -32,9 +33,21 @@
 
 class ConnectionProviderTest {
 
+	static final ConnectionProvider.MeterRegistrar METER_REGISTRY = new ConnectionProvider.MeterRegistrar() {
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/ConnectionProviderTest.java:36]: Once we have the new method with a default implementation we can revert the changes in this class.

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 3]
```diff
@@ -69,36 +70,47 @@ void tearDown() throws Exception {
...
+		triggerAcquisition(false, () -> new MeterRegistrarImpl(registered, null, null));
+		assertThat(registered.get()).isFalse();
+	}
+
+	@Test
+	void disposeWhenMetricsDeregistered() {
+		AtomicBoolean registered = new AtomicBoolean();
+		AtomicBoolean deRegistered = new AtomicBoolean();
+
+		Connection conn = triggerAcquisition(true, () -> new MeterRegistrarImpl(registered, deRegistered, null));
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:105]:
```suggestion
		triggerAcquisition(true, () -> new MeterRegistrarImpl(registered, deRegistered, null));
```

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 4]
```diff
@@ -69,36 +70,47 @@ void tearDown() throws Exception {
...
+		assertThat(registered.get()).isTrue();
+		assertThat(deRegistered.get()).isFalse();
+
+		pool.disposeWhen(remoteAddress.get());
+		assertThat(deRegistered.get()).isTrue();
 	}
 
-	private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvider.MeterRegistrar> registrarSupplier) {
+	private Connection triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvider.MeterRegistrar> registrarSupplier) {
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:113]: Connection establishment is expected to fail always so we cannot return any Connection. The intention of this method is just to create the connection pool.

```
io.netty.channel.AbstractChannel$AnnotatedSocketException: Can't assign requested address: localhost/127.0.0.1:0
Caused by: java.net.BindException: Can't assign requested address
	at sun.nio.ch.Net.connect0(Native Method)
	at sun.nio.ch.Net.connect(Net.java:482)
	at sun.nio.ch.Net.connect(Net.java:474)
```

```suggestion
	private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvider.MeterRegistrar> registrarSupplier) {
```

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 5]
```diff
@@ -108,16 +120,46 @@ private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvi
 		ClientTransportConfigImpl config =
 				new ClientTransportConfigImpl(group, pool, Collections.emptyMap(), remoteAddress);
 
+		Connection conn = null;
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:123]:
```suggestion
```

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 6]
```diff
@@ -108,16 +120,46 @@ private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvi
...
 		try {
-			pool.acquire(config, ConnectionObserver.emptyListener(), remoteAddress, config.resolverInternal())
+			conn = pool.acquire(config, ConnectionObserver.emptyListener(), remoteAddress, config.resolverInternal())
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:125]:
```suggestion
			pool.acquire(config, ConnectionObserver.emptyListener(), remoteAddress, config.resolverInternal())
```

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 7]
```diff
@@ -108,16 +120,46 @@ private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvi
...
 			    .block(Duration.ofSeconds(10L));
 			fail("Exception is expected");
 		}
 		catch (Exception expected) {
 			// ignore
 		}
+		return conn;
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:132]:
```suggestion
```

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 8]
```diff
@@ -108,16 +120,46 @@ private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvi
...
+	static final class MeterRegistrarImpl implements ConnectionProvider.MeterRegistrar {
+		AtomicBoolean registered;
+		AtomicBoolean deRegistered;
+		AtomicInteger customMetric;
+
+		MeterRegistrarImpl(AtomicBoolean registered, AtomicBoolean deRegistered, AtomicInteger customMetric) {
+			this.registered = registered;
+			this.deRegistered = deRegistered;
+			this.customMetric = customMetric;
+		}
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:144]: Let's announce that `Nullable` is allowed.

```suggestion
		MeterRegistrarImpl(
				@Nullable AtomicBoolean registered,
				@Nullable AtomicBoolean deRegistered,
				@Nullable AtomicInteger customMetric) {
			this.registered = registered;
			this.deRegistered = deRegistered;
			this.customMetric = customMetric;
		}

```

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 9]
```diff
@@ -108,16 +120,46 @@ private void triggerAcquisition(boolean metricsEnabled, Supplier<ConnectionProvi
...
+			}
+		}
+
+		@Override
+		public void deRegisterMetrics(String poolName, String id, SocketAddress remoteAddress) {
+			if (deRegistered != null) {
+				deRegistered.set(true);
+			}
+		}
+	};
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:161]:
```suggestion
	}
```

File: reactor-netty-http/src/main/java/reactor/netty/http/client/MicrometerHttp2ConnectionProviderMeterRegistrar.java [hunk 10]
```diff
@@ -59,4 +60,13 @@ void registerMetrics(String poolName, String id, SocketAddress remoteAddress, In
 		     .tags(tags)
 		     .register(REGISTRY);
 	}
+
+	void deRegisterMetrics(String poolName, String id, SocketAddress remoteAddress) {
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/MicrometerHttp2ConnectionProviderMeterRegistrar.java:64]: In `Http2ConnectionProvider` we need to override `reactor.netty.resources.PooledConnectionProvider#deRegisterDefaultMetrics`

```
	@Override
	protected void deRegisterDefaultMetrics(String id, SocketAddress remoteAddress) {
		MicrometerHttp2ConnectionProviderMeterRegistrar.INSTANCE
				.deRegisterMetrics(name(), id, remoteAddress);
	}
```

File: reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java [hunk 11]
```diff
@@ -34,6 +34,7 @@
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import reactor.netty.Connection;
```
Comment (violetagg) [inline reactor-netty-core/src/test/java/reactor/netty/resources/PooledConnectionProviderCustomMetricsTest.java:37]:
```suggestion
```

**Review by violetagg — CHANGES_REQUESTED** [review]
@manolama Thanks for the PR!

I have several comments.

Also I want to add the link below for completeness
https://micrometer.io/docs/concepts#_why_is_my_gauge_reporting_nan_or_disappearing

Once Reactor Netty disposes a connection pool there is no strong reference anymore, so NaN is expected to be returned by the Gauges.

File: reactor-netty-core/src/main/java/reactor/netty/resources/ConnectionProvider.java (hunk 1 above)
Comment (manolama) [reply reactor-netty-core/src/main/java/reactor/netty/resources/ConnectionProvider.java:820]: Sounds good, thanks!

**Comment by manolama** [comment]
Thanks!

> Once Reactor Netty disposes a connection pool there is no strong reference anymore, so NaN is expected to be returned by the Gauges.
Right, that works as expected. Unfortunately the Meter and Id stick around in the Micrometer map so we want to remove those to release the Id strings and stop reporting.

Context for the PR (from internal chats):
We have some services that exhibit a mesh behavior where an app may connect to multiple, individual instances of other server apps in a group. A connection pool is instantiated for each of those instances. We also have frequent deployments of the server apps and each new instance picks up a different IP. So if a client app is long lived and multiple deployments occur for the server apps, we wind up with stale Meters in Micrometer. This will help remove stale Meters.

**Review by violetagg — APPROVED** [review]

**Comment by violetagg** [comment]
@reactor/netty-team PTAL

the failing tests on Windows OS are not related
the build on Mac OS with native transport seems to be hanging but I don't think it is related

File: reactor-netty-core/src/test/java/reactor/netty/resources/ConnectionProviderTest.java [hunk 12]
```diff
@@ -127,4 +127,4 @@ public int permitMaximum() {
 		public void returnPermits(int returned) {
 		}
 	}
-}
+}
```
Comment (pderop) [inline reactor-netty-core/src/test/java/reactor/netty/resources/ConnectionProviderTest.java:130]: nit: the newline has been removed at the end of the file, can you just re-add it, so the ConnectionProviderTest won't appear in the list of changes ?

Comment (manolama) [reply reactor-netty-core/src/test/java/reactor/netty/resources/ConnectionProviderTest.java]: absolutely :), sorry.

**Comment by violetagg** [comment]
@manolama I back ported this change to 1.0.x for 1.0.26

File: reactor-netty-core/src/main/java/reactor/netty/resources/ConnectionProvider.java (hunk 1 above)
Comment (violetagg) [reply reactor-netty-core/src/main/java/reactor/netty/resources/ConnectionProvider.java]: I removed the default implementation in Reactor Netty 2 with PR #2613
